src/pipelinedefinition.cpp

```diff
@@ -85,14 +87,76 @@ Status PipelineDefinition::validate(ModelManager& manager) {
+    for (const auto& nodeInfo : nodeInfos) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            void* customNodeLibraryInternalManager = nullptr;
+            auto it = nodeResources.find(nodeInfo.nodeName);
+            bool internalManagerExists = (it != nodeResources.end());
+            if (internalManagerExists) {
+                customNodeLibraryInternalManager = nodeResources.at(nodeInfo.nodeName);
+            }
+            auto params = createCustomNodeParamArray(nodeInfo.parameters).get();
+            if (!nodeInfo.library.isValid()) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Pipeline: {} node: {} refers to incorrect library", pipelineName, nodeInfo.nodeName);
+                return StatusCode::PIPELINE_DEFINITION_INVALID_NODE_LIBRARY;
+            }
+            auto status = nodeInfo.library.initialize(&customNodeLibraryInternalManager, params, nodeInfo.parameters.size());
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Initialization of library with base path: {} failed", nodeInfo.library.basePath);
+                return StatusCode::NODE_LIBRARY_INITIALIZE_FAILED;
+            }
+            if (!internalManagerExists) {
+                nodeResources.insert({nodeInfo.nodeName, customNodeLibraryInternalManager});
+            }
+        }
+    }
+    return StatusCode::OK;
+}
+
+void PipelineDefinition::deinitializeUnusedNodeResources(const std::vector<NodeInfo>& nodeInfos) {
+    std::vector<NodeInfo> diff;
+    for (const auto& nodeInfo : this->nodeInfos) {
+        auto it = std::find_if(nodeInfos.begin(), nodeInfos.end(),
+            [&nodeInfo](const auto& x) { return x.nodeName == nodeInfo.nodeName; });
+        if (it == nodeInfos.end()) {
+            diff.push_back(nodeInfo);
+        }
+    }
+
+    for (const auto& nodeInfo : diff) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            void* customNodeLibraryInternalManager = nodeResources.at(nodeInfo.nodeName);
+            auto status = nodeInfo.library.deinitialize(customNodeLibraryInternalManager);
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Deinitialization of library with base path: {} failed", nodeInfo.library.basePath);
+            }
+            nodeResources.erase(nodeInfo.nodeName);
+        }
+    }
+}
+
+void PipelineDefinition::deinitializeNodeResources() {
+    for (const auto& nodeInfo : this->nodeInfos) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            void* customNodeLibraryInternalManager = nodeResources.at(nodeInfo.nodeName);
+            auto status = nodeInfo.library.deinitialize(customNodeLibraryInternalManager);
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Deinitialization of library with base path: {} failed", nodeInfo.library.basePath);
+            }
+        }
+    }
+    nodeResources.clear();
+}
```

Comment:
I think it would be better to have:
1) Function that call deinitialize on set of libraries & removes them from nodeResources
2) Function that calculates the diff (set of now unused libraries)